File: internal/voice/gateway/frames.go
// Package gateway implements the WebSocket endpoint at
// /api/v1/voice/sessions. It mux-es the 8 client→gateway frames and the 10
// gateway→client frames specified in DES-0009 §2 — the names below are
// reproduced verbatim and are the load-bearing public contract every web /
// mobile client subscribes to.
package gateway

// Client → gateway frame types (DES-0009 §2 — 8 frames, last one is keepalive).
const (
	ClientFrameStartAudio     = "start_audio"
	ClientFrameStopAudio      = "stop_audio"
	ClientFrameMute           = "mute"
	ClientFrameUnmute         = "unmute"
	ClientFrameCancelSpeech   = "cancel_speech"
	ClientFrameLanguageSwitch = "language_switch"
	ClientFrameQuickAction    = "quick_action"
	ClientFramePing           = "ping"
)

// ClientFrameNames is the canonical list — tests and admin tooling enumerate
// this. Order matches DES-0009 §2.
var ClientFrameNames = []string{
	ClientFrameStartAudio,
	ClientFrameStopAudio,
	ClientFrameMute,
	ClientFrameUnmute,
	ClientFrameCancelSpeech,
	ClientFrameLanguageSwitch,
	ClientFrameQuickAction,
	ClientFramePing,
}

// Gateway → client frame types (DES-0009 §2 — 10 frames).
const (
	GatewayFrameSessionOpen      = "session_open"
	GatewayFrameSessionClosed    = "session_closed"
	GatewayFrameTranscript       = "transcript"
	GatewayFrameFunctionCall     = "function_call"
	GatewayFrameFunctionResult   = "function_result"
	GatewayFrameAudioMeter       = "audio_meter"
	GatewayFrameProviderSwitched = "provider_switched"
	GatewayFramePong             = "pong"
	GatewayFrameError            = "error"
	GatewayFrameQuotaWarning     = "quota_warning"
)

// GatewayFrameNames is the canonical list. Order matches DES-0009 §2.
var GatewayFrameNames = []string{
	GatewayFrameSessionOpen,
	GatewayFrameSessionClosed,
	GatewayFrameTranscript,
	GatewayFrameFunctionCall,
	GatewayFrameFunctionResult,
	GatewayFrameAudioMeter,
	GatewayFrameProviderSwitched,
	GatewayFramePong,
	GatewayFrameError,
	GatewayFrameQuotaWarning,
}

// ClientFrame is the JSON envelope from a client.
type ClientFrame struct {
	Type    string         `json:"type"`
	Payload map[string]any `json:"payload,omitempty"`
}

// GatewayFrame is the JSON envelope to a client.
type GatewayFrame struct {
	Type    string         `json:"type"`
	Payload map[string]any `json:"payload,omitempty"`
}

// QuickActionName enumerates the four quick actions a kiosk button can fire
// (REQ-0011 AC-8, DES-0008 §6).
const (
	QuickActionRecommendations = "recommendations"
	QuickActionPromotions      = "promotions"
	QuickActionAllergy         = "allergy"
	QuickActionHelp            = "help"
)

File: apps/kiosk/lib/voice/frames.ts
// DES-0009 §2 WebSocket control frame schema.

// ───── Client → gateway (8 frames; `ping` is keepalive) ─────
export type ClientFrame =
  | { type: "start_audio" }
  | { type: "stop_audio" }
  | { type: "mute" }
  | { type: "unmute" }
  | { type: "cancel_speech" }
  | { type: "language_switch"; bcp47: string }
  | { type: "quick_action"; name: "recommendations" | "promotions" | "allergy" | "help" }
  | { type: "ping" };

export const CLIENT_FRAME_TYPES = [
  "start_audio",
  "stop_audio",
  "mute",
  "unmute",
  "cancel_speech",
  "language_switch",
  "quick_action",
  "ping",
] as const;

// ───── Gateway → client (10 frames) ─────
export type ServerFrame =
  | { type: "session_open"; session_id: string; provider: string }
  | { type: "session_closed"; reason: string }
  | {
      type: "transcript";
      role: "user" | "assistant";
      text: string;
      final: boolean;
      interrupted?: boolean;
    }
  | { type: "function_call"; name: string; arguments_json: string; call_id: string }
  | {
      type: "function_result";
      call_id: string;
      success: boolean;
      payload?: unknown;
      error?: string;
    }
  | { type: "audio_meter"; db: number }
  | { type: "provider_switched"; from: string; to: string; reason: string }
  | { type: "pong" }
  | { type: "error"; code: string; message: string }
  | { type: "quota_warning"; tenant_id: string; used_minutes: number; cap_minutes: number };

export const SERVER_FRAME_TYPES = [
  "session_open",
  "session_closed",
  "transcript",
  "function_call",
  "function_result",
  "audio_meter",
  "provider_switched",
  "pong",
  "error",
  "quota_warning",
] as const;
